src/python.c

```diff
@@ -1642,6 +1685,51 @@ canSort(MulticornPlanState * state, List *deparsed)
 	return result;
 }
 
+/*
+ * Call the can_pushdown_upperrel method from the python implementation, to
+ * determine whether upper relations can be pushed down to the corresponding
+ * data source to begin with.
+ *
+ * If yes, then also initialize some fields in MulticornPlanState needed for
+ * more granular conditional logic for assesing whether the particular query
+ * is suitable for pushdown.
+ */
+bool
+canPushdownUpperrel(MulticornPlanState * state)
+{
+    PyObject    *fdw_instance = state->fdw_instance,
+                *p_upperrel_pushdown,
+                *p_object;
+    bool pushdown_upperrel = false;
+
+    p_upperrel_pushdown = PyObject_CallMethod(fdw_instance, "can_pushdown_upperrel", "()");
+    errorCheck();
+
+    if (p_upperrel_pushdown != NULL && p_upperrel_pushdown != Py_None)
+    {
+        /* Determine whether the FDW instance supports GROUP BYs */
+        p_object = PyMapping_GetItemString(p_upperrel_pushdown, "groupby_supported");
+        if (p_object != NULL && p_object != Py_None)
+		{
+            state->groupby_supported = PyObject_IsTrue(p_object);
+            Py_DECREF(p_object);
+		}
+
+        /* Determine which aggregation functions are supported */
+        p_object = PyMapping_GetItemString(p_upperrel_pushdown, "agg_functions");
+        if (p_object != NULL && p_object != Py_None)
+		{
+            state->agg_functions = PyMapping_Keys(p_object);
+            Py_DECREF(p_object);
+		}
+
+        pushdown_upperrel = true;
+    }
+
+	Py_DECREF(p_upperrel_pushdown);
```

Comment:
Should this DECREF be inside of the ` if (p_upperrel_pushdown != NULL && p_upperrel_pushdown != Py_None)` like other decrefs?

Reply:
We'd need to decrement a `Py_None` reference, which would leak if the `Py_DECREF` was in the if statement I believe.

That said, in the case of `p_upperrel_pushdown` being a null pointer it seems the [proper approach](https://stackoverflow.com/questions/41915305/is-there-any-benefit-to-using-py-decref-instead-of-py-xdecref-for-python-c-exten) is to use `Py_XDECREF`, like in the case of `pythonDictToTuple` function.

Also seems like I should do something similar for `p_object` inside the outer if statement.

Adding those changes now.